### ISSUE TITLE
feat: API for inhomogeneous (non-zero) Dirichlet boundary conditions

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/finitevolX#186]
## Summary

Add first-class API support for **inhomogeneous Dirichlet boundary conditions** — solving $(A - \lambda)\psi = f$ with $\psi = g$ on the boundary, where $g \neq 0$.  Currently, all solvers (spectral, capacitance, CG, multigrid) assume homogeneous BCs ($\psi = 0$).  Users must manually implement the lifting trick, which is error-prone (see the `demo_solvers` notebook Section 8 for a worked example).

## Motivation

Non-zero boundary values arise in many practical settings:

| Use case | Boundary data source |
|----------|---------------------|
| **Regional nesting** | Parent model provides SSH/streamfunction at open boundaries |
| **Data assimilation** | Tide gauges or satellite altimetry fix boundary values |
| **Reanalysis-forced runs** | ERA5 or ORAS5 SSH prescribed at lateral boundaries |
| **Two-way coupling** | Outer model provides BCs for an inner high-resolution domain |
| **Manufactured solutions** | Testing with non-trivial exact solutions that don't vanish at boundaries |

The lifting trick (decompose $\psi = \psi_{\text{lift}} + \psi_{\text{hom}}$, correct the RHS, solve the homogeneous problem) is mathematically straightforward but operationally fiddly:

1. Users must identify the correct "boundary cells" (dry cells adjacent to wet, not the coast ring itself)
2. The lifting function must be constructed carefully (values at dry boundary cells, zero in interior)
3. The corrected RHS $f' = f - A(\psi_{\text{lift}})$ requires calling the operator on the lift
4. The final reconstruction $\psi = \psi_{\text{lift}} + \psi_{\text{hom}}$ is easy to forget

An API that handles this transparently would make the library much more practical for realistic applications.

### Option A: `bc_values` parameter on convenience wrappers

The simplest approach — add an optional `bc_values` parameter to the existing convenience functions:

```python
# Current API (homogeneous only)
psi = fvx.streamfunction_from_vorticity(zeta, dx, dy, bc="dst")

# Proposed: pass boundary values directly
psi = fvx.streamfunction_from_vorticity(
    zeta, dx, dy, bc="dst",
    bc_values=g,  # array of shape (Ny, Nx), nonzero at boundary cells
)

# Same for pressure and PV inversion
p = fvx.pressure_from_divergence(
    div_u, dx, dy, bc="dct",
    bc_values=p_boundary,
)
psi = fvx.pv_inversion(
    pv, dx, dy, lambda_=1/Ld**2,
    bc_values=psi_boundary,
)
```

**Internally**, when `bc_values` is not None, the wrapper:
1. Builds the lift: `psi_lift = bc_values * (~mask)` (nonzero only at dry cells)
2. Corrects RHS: `rhs_corrected = rhs - A(psi_lift)`
3. Solves homogeneous: `psi_hom = solver(rhs_corrected)`
4. Returns: `psi_lift + psi_hom`

### Option B: `InhomogeneousBCSolver` wrapper class

An equinox Module that wraps any homogeneous solver and adds the lifting logic:

```python
from finitevolx import InhomogeneousBCSolver

# Wrap any existing solver
base_solver = fvx.build_capacitance_solver(mask, dx, dy, lambda_=0.0)
solver = InhomogeneousBCSolver(
    base_solver=base_solver,
    mask=mask,
    dx=dx, dy=dy,
    lambda_=0.0,
)

# Solve with prescribed boundary values
psi = solver(rhs, bc_values=g)

# Also works with CG
solver_cg = InhomogeneousBCSolver(
    base_solver=lambda rhs: fvx.solve_cg(A, rhs, preconditioner=pc)[0],
    mask=mask,
    dx=dx, dy=dy,
    lambda_=4.0,
)
psi = solver_cg(rhs, bc_values=g)
```

### Option C: Standalone `apply_inhomogeneous_bc` function

A pure function that does the lifting trick, composable with any solver:

```python
from finitevolx import apply_inhomogeneous_bc

# Step 1: correct the RHS for non-zero boundary data
rhs_corrected, psi_lift = apply_inhomogeneous_bc(
    rhs=rhs,
    bc_values=g,          # prescribed values at boundary
    mask=mask,             # wet/dry mask
    dx=dx, dy=dy,
    lambda_=lambda_,
)

# Step 2: solve the corrected (homogeneous) problem with ANY solver
psi_hom = my_favorite_solver(rhs_corrected)

# Step 3: reconstruct
psi = psi_lift + psi_hom
```

This is the most flexible — it doesn't couple to any specific solver.

### Option D: `BoundaryCondition` object (most general)

For future extensibility (Robin, mixed, open BCs), define a BC object:

```python
from finitevolx import DirichletBC, NeumannBC

# Dirichlet with known values
bc = DirichletBC(values=g, mask=mask)

# Neumann with known flux
bc = NeumannBC(flux=dg_dn, mask=mask)

# Pass to any solver
psi = fvx.streamfunction_from_vorticity(zeta, dx, dy, bc=bc)
```

This is the most forward-looking but the most complex to implement.

## Recommended approach

**Start with Option C** (`apply_inhomogeneous_bc`) as it's:
- Simple to implement (pure function, ~30 lines)
- Composable with all existing solvers (spectral, CG, capacitance, multigrid)
- Easy to test (just verify the RHS correction is correct)
- No changes to existing solver signatures

Then **add Option A** (bc_values parameter on convenience wrappers) that calls Option C internally — this gives users the one-line API for common cases.

### Identifying boundary cells

The "boundary cells" for the lifting function are the **dry cells adjacent to at least one wet cell** — i.e., the 1-cell-wide land ring around the ocean:

```python
from scipy.ndimage import binary_dilation

wet = mask.astype(bool)
dry_boundary = binary_dilation(wet) & ~wet
```

For the C-grid ghost-cell convention, this corresponds to the ghost cells that the 5-point Laplacian stencil reads when computing values at boundary-adjacent wet cells.

### The lifting function

```python
def apply_inhomogeneous_bc(rhs, bc_values, mask, dx, dy, lambda_=0.0):
    """Correct RHS for inhomogeneous Dirichlet BCs via the lifting trick.

    Parameters
    ----------
    rhs : array (Ny, Nx)
        Original right-hand side (masked: zero at dry cells).
    bc_values : array (Ny, Nx)
        Prescribed boundary values.  Only values at dry cells adjacent
        to the wet domain are used; interior values are ignored.
    mask : array (Ny, Nx)
        Binary wet/dry mask (1=wet, 0=dry).
    dx, dy : float
        Grid spacings.
    lambda_ : float
        Helmholtz parameter.

    Returns
    -------
    rhs_corrected : array (Ny, Nx)
        Corrected RHS for the homogeneous problem.
    psi_lift : array (Ny, Nx)
        Lifting function (add to homogeneous solution to get full solution).
    """
    wet = mask.astype(bool)
    dry_boundary = binary_dilation(wet) & ~wet

    # Build lift: prescribed values at dry boundary, zero elsewhere
    psi_lift = jnp.where(dry_boundary, bc_values, 0.0)

    # Correct RHS: subtract the Laplacian of the lift at wet cells
    A_lift = masked_laplacian(psi_lift, mask, dx, dy, lambda_=lambda_)
    rhs_corrected = rhs - A_lift

    return rhs_corrected, psi_lift
```

### Reconstruction

```python
psi_hom = solver(rhs_corrected)
psi = psi_lift + psi_hom  # full solution with non-zero BCs
```

### Edge cases

- **Spectral solvers** (DST/DCT/FFT): These operate on the interior grid without a mask. The lift function would need to be handled differently — the ghost-cell values in the C-grid array encode the BCs. For DST with ghost=0, inhomogeneous BCs mean setting the ghost cells to g instead of 0 before extracting the interior. This may need a separate code path.
- **Capacitance solver**: The `CapacitanceSolver` from spectraldiffx enforces ψ=0 at inner-boundary points. For inhomogeneous BCs, the correction term propagates through the Green's function. The lifting trick works as-is.
- **Time-dependent problems**: If boundary values change every time step (e.g., time-varying reanalysis), the correction must be recomputed at each step. The function should be JIT-friendly for this use case.

## Testing strategy

1. **Manufactured solution**: Choose $\psi_{\text{exact}} = \sin(\pi x)\sin(\pi y) + 0.1\sin(2\pi y)$ (the second term doesn't vanish at boundaries). Compute $f = (A-\lambda)\psi_{\text{exact}}$, extract $g = \psi_{\text{exact}}|_{\text{boundary}}$, solve with inhomogeneous BCs, verify $\|\psi - \psi_{\text{exact}}\| < \epsilon$.
2. **Consistency check**: Homogeneous BCs (g=0) should give the same result as the current API.
3. **All solver types**: Test with spectral, CG, capacitance, and multigrid.
4. **Masked domains**: Test on basin, circle, and irregular masks.

## Acceptance criteria

- [ ] `apply_inhomogeneous_bc(rhs, bc_values, mask, dx, dy, lambda_)` function implemented
- [ ] Convenience wrappers accept `bc_values` parameter
- [ ] Works with all solver methods (spectral, CG, capacitance, multigrid)
- [ ] Manufactured-solution test passes to solver tolerance
- [ ] JIT-compatible for time-stepping applications
- [ ] Documentation in elliptic solver guide
- [ ] Example in `demo_solvers` notebook (already has a manual version in Section 8)

## Labels

`enhancement`, `solvers`
